### DISCUSSION TITLE
render in wireframe

### DISCUSSION THREAD

**Discussion by phatpixels** [google/model-viewer#4336] [Q&A]
Hi All,

Is there a way to display the model in simple wireframe view? say switch modes from fully textured to simple wireframe to see model construction. Thanks.

**Comment by WadeMorrison** [comment]
Hey there! I think to get the desired result you'll need to use the addPass() in the [Custom Effects](https://modelviewer.dev/examples/postprocessing/#custom-effects) section of the Post Processing Effects. Hope that helps!

**Reply by phatpixels** [reply]
Thanks for the reply @WadeMorrison, I could be wrong but isn't that just a grid shader not the wireframe view that I was after? I'm not after an effect but the real wireframe view of the 3d model is that possible in M-V?

**Comment by NicoPennec** [comment]
Hi @phatpixels !

**Three.js** has a `wireframe` option on the [MeshStandardMaterial](https://threejs.org/docs/#api/en/materials/MeshStandardMaterial.wireframe).
This feature is not supported out-of-the-box by **Model-Viewer**, but you can access it as explained in the [FAQ](https://modelviewer.dev/docs/faq.html#entrydocs-general-questions-three).

I just added a wireframe rendering mode to my current project (see the [source code](https://github.com/cgwire/kitsu/blob/cc674a0e33804883a0b02088bb4c7bd9020de73a/src/components/previews/ObjectViewer.vue#L67-L92) of Kitsu)

To reach this goal, I create a new material variant from my 3D model (see [`createMaterialInstanceForVariant(...)`](https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/features/scene-graph/model.ts#L295) )
Then I enable the `wireframe` option to each material of my variant, and I apply the gray color to the wire (but it's optional).
Finally I swap to the variant in Model-Viewer (see [doc](https://modelviewer.dev/examples/scenegraph/#variants))

The result:
<img width="535" alt="wireframe" src="https://github.com/google/model-viewer/assets/493223/6bbd4d67-3294-43b9-8886-b1e07cf0f406">

**Comment by core-hacked** [comment]
Edit: I was just stupid, you can hook it and use scene.traverse as you would with threejs. Sorry..
```js
var mvElement = document.getElementById("mv-element");
              var scene =
                mvElement[
                  Object.getOwnPropertySymbols(mvElement).find(
                    (e) => e.description === "scene"
                  )
                ];

              // traverse the scene and set wireframe to true
              scene.traverse(function (child) {
                if (child.isMesh) {
                  child.material.wireframe = true;
                  child.material.color.setHex(0xff0000); // sets color
                }
              });
```


> Hi @ phatpixels !
> 
> **Three.js** has a `wireframe` option on the [MeshStandardMaterial](https://threejs.org/docs/#api/en/materials/MeshStandardMaterial.wireframe). This feature is not supported out-of-the-box by **Model-Viewer**, but you can access it as explained in the [FAQ](https://modelviewer.dev/docs/faq.html#entrydocs-general-questions-three).
> 
> I just added a wireframe rendering mode to my current project (see the [source code](https://github.com/cgwire/kitsu/blob/cc674a0e33804883a0b02088bb4c7bd9020de73a/src/components/previews/ObjectViewer.vue#L67-L92) of Kitsu)
> 
> To reach this goal, I create a new material variant from my 3D model (see [`createMaterialInstanceForVariant(...)`](https://github.com/google/model-viewer/blob/master/packages/model-viewer/src/features/scene-graph/model.ts#L295) ) Then I enable the `wireframe` option to each material of my variant, and I apply the gray color to the wire (but it's optional). Finally I swap to the variant in Model-Viewer (see [doc](https://modelviewer.dev/examples/scenegraph/#variants))
> 
> The result: <img alt="wireframe" width="535" src="https://private-user-images.githubusercontent.com/493223/273265502-6bbd4d67-3294-43b9-8886-b1e07cf0f406.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwODk4MTQsIm5iZiI6MTcwNTA4OTUxNCwicGF0aCI6Ii80OTMyMjMvMjczMjY1NTAyLTZiYmQ0ZDY3LTMyOTQtNDNiOS04ODg2LWIxZTA3Y2YwZjQwNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMlQxOTU4MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01Yzk2NmZkODkyMzZlOTIzMzgwMDU5NjhhMTQyZjhkY2Y5ODgyMGFiMTY4NDhiYWFjMmQ5ZjA5MmJjYjUzZTMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kTOhDs4ZQmWNMJDcgtETaGc5rS9N7DvtqGdqHNOt1Bs">

Hi, I was wondering if you could provide the code example of that, as either I am too stupid to figure it out or I am missinterpreting what you wrote.

**Reply by Sembiance** [reply]
This works great! Thanks! One tip, the wireframe won't switch until the user interacts with the model (or if it's already animating). So to get it to switch to wireframe right away and re-draw, call: mvElement.updateFraming()

**Reply by elalish** [reply]
Even better is `scene.queueRender();`

**Comment by RomainMazB** [comment]
I started with @NicoPennec answers but figured out that while it was working as expected at the first glance.
If you look the wires colors in details, they are affected by the texture:
![pumpkin-wireframe](https://github.com/user-attachments/assets/bda41d62-81eb-47b2-92fa-d73c9c2ed852)

Then I discovered that you can basically disable the material texture using [the map option of Three.js](https://threejs.org/docs/#api/en/materials/MeshStandardMaterial.wireframe) .

Here is the final code to create the wireframe variant, based on [the source code of Kitsu already linked above](https://github.com/cgwire/kitsu/blob/cc674a0e33804883a0b02088bb4c7bd9020de73a/src/components/previews/ObjectViewer.vue#L67-L92):
```js
createWireframeVariant(model) {
    const maxIndex = model.materials.length
    for (let i = 0; i < maxIndex; i++) {
        const variantMaterial = model.createMaterialInstanceForVariant(
            i,
            `material-wireframe-${i}`,
            'variant-wireframe',
            this.isWireframe
        )
        if (!variantMaterial) {
            continue
        }
        const texture = variantMaterial.normalTexture
        const materialsSymbol = Object.getOwnPropertySymbols(texture).find(
            symbol => symbol.description === 'materials'
        )
        const materials = texture[materialsSymbol]
        materials.forEach(material => {
            material.wireframe = true // Enable the wireframe
            material.map = false // Disable the texture
        })
    }
}
```
The final result:
![blank-wireframe](https://github.com/user-attachments/assets/93c1f76f-8ea1-45bb-abe8-16bfb88e636d)

**Comment by IorHon** [comment]
Its posible to see only the edges? Like CAD view?
